Updated serialization documentation, refactor some serialization upgrade scripts (#5022)
oscar-system/Oscar.jl: docs/src/General/serialization.md

TropicalHypersurface
```

## Container Types

> The container types used for storing and organizing data of the types listed above can also be serialized. 

Reviewer: But what are "the container types" referenced here? Shouldn't this be stated here?

> There are some caveats, for example when constructing a `Vector` with entries of different types, julia will use `typejoin` to find the closest common ancestor among the types.

Reviewer: "for example"? Shouldn't the point of the documentation be to list *all* the caveats?

> This may result in a type that can no longer be serialized as a `Vector`, this is due to the fact that the type parameters for serialization are not necessarily the type parameters for the data. 

Reviewer: > This may result in a type that can no longer be serialized as a `Vector`, [...]

Multiple remarks:
- Instead of "This may result in a type" perhaps better say "This may result in a container"
- the sentence sounds as if there *is* a way to serialize these, just not as a "Vector".
- in the end, this is a somewhat arbitrary design restriction on our side (and one we could surely lift if we wanted to). So how about:

> This may result in a container that our serialization framework currently is not able to serialize. [...]

Finally, does this language mean that there may be `Vector` which are not homogeneous (i.e. not all members have identical type) but which are nevertheless serializable? What would be an example?

Author: there are ways to serialize but not as a vector. In the doc string below the error says to use a tuple instead.

The thing is we will not be able to serialize non homogeneous vectors  in the future either. Serializing Vectors means each entry needs to have the same "type parameters".

What I am trying to say here, is that julia might show that the vector is homogeneous because all entries are of the same type, but that doesn't imply homogeneous in the Oscar serialization sense.

Author: please see updated text and let me know what you think.

To serialize a `Vector` all entries must have the output of [`type_params`](@ref) be equal (`==` returns true), this allows us to provide a size efficient storage of homogeneous data. The `Tuple` and `NamedTuple` types will store type information (type names and parameters) for each entry. 
Storing `Dict` types depends on the key and value types, for example if the keys are either `String`, `Symbol` or `Int` then the type parameters for the value may vary by key. 

> In the other cases the type parameters for the keys and the values must be the same for all keys and values.

Reviewer: Wait, so do you mean that if `(k,v)` is a key-value pair, then `type_params(k) == type_params(v)` must hold? Or do you mean
```suggestion
In the other cases the type parameters for all keys must be equal, and likewise all values must have equal type parameters."
```

> ```

Reviewer: Shouldn't this be a doctest?
````suggestion
```jldoctest
````

Author: I could make it a doc test but then I need a temporary file path, and I would need to display the whole error?

Reviewer: Errors can be abbreviated, see https://documenter.juliadocs.org/stable/man/doctests/#Exceptions

Regarding the temporary file path, just use `/tmp/file.mrdi` here -- due to the error it won't be used anyway?

julia> Qxy, (x, y) = QQ[:x, :y]
(Multivariate polynomial ring in 2 variables over QQ, QQMPolyRingElem[x, y])

julia> Qt, t = QQ[:t]
(Univariate polynomial ring in t over QQ, t)

julia> typeof([t, x, y])
Vector{RingElem} (alias for Array{RingElem, 1})

julia> save("/path/to/file.mrdi", [t, x, y])
ERROR: ArgumentError: Not all type parameters of Vector or Matrix entries are the same, consider using a Tuple for serialization

> 
> ...

Reviewer: ```suggestion
```

```

> ## Type Parameters

Reviewer: The section about "containers" before this already heavily references "type parameters", wouldn't it be better to swap them, and explain the type parameters first?

> The `TypeParams` type is used internally, and is a container struct for the type and the parameters need to serialize an object. Since it can be useful for users to create them when overriding parmaeters on load we include a small section here.

Reviewer: Typos, split to multiple lines
```suggestion
The `TypeParams` type is used internally, and is a container struct for the
type and the parameters needed to serialize an object. Since it can be useful
for users to create them when overriding parameters on load we include a small
section here.
```

> The `TypeParams` type is used internally, and is a container struct for the type and the parameters need to serialize an object. Since it can be useful for users to create them when overriding parmaeters on load we include a small section here.

Reviewer: "is a container struct" -> I find this phrasing *highly* confusing given the topic of the previous section. 

Also this text refers to "type and parameters needed to serialize an object". Is "type" simply referring to the Julia type of the object? Then this should be explicitly stated. Also what are these "parameters" -- where is that explained? Shouldn't it be explained first? If that is not possible due to circular dependencies, can we at least get a `@ref` between the relevant sections?

```

```@repl oscar
Oscar.Serialization.reverse_type_map
```

## Type Parameters

> The `TypeParams` type is used internally, and is a struct for the julia type and the parameters needed to serialize an object. 

Reviewer: > [...] and is a struct for the julia type and the parameters [...]

I find this ambiguous. Does that mean `TypeParams` exists "for" the sake of the julia type so it can be used? Or is it meant "for" encoding the julia type and the parameters? Or something else?

(I would try to make a constructive concrete suggestion for rephrasing this but I am genuinely uncertain what is intended. Perhaps it would make sense if we met and set together for a bit to hash out the text together)

> The parameters needed to serialize OSCAR types are not necessarily the same as the parameters of the type itself.

Reviewer: This may sound like a stupid nitpick, but I honestly think this is a potential source of confusion: we are not serializing types, we are serializing *instances* of types. And these parameters are not "parameters" of the *types* but rather of those instances...

If this observation is followed, the obvious next step is to rename `TypeParam`. Then we would not have to explain at length how "OSCAR serialization type parameters" are not the same as "Julia type parameters"...

But it seems there is more to this than "type" vs. "instance of type" ... as far as I can tell, you are using this not for all types/instances, but rather for "element types" / "elements", i.e. for things with parents ?

Author: @fingolfin what about `EncodingParams` ?

> For example, take `p` with julia type `MPolyRingElem{QQField}`, this julia type is parametrized by the type of its coefficients, however

Reviewer: ```suggestion
For example, consider a Julia object `p` with Julia type `MPolyRingElem{QQField}`.
This julia type is parametrized by the type of its coefficients, however
```

the parameter needed for serializing is its parent, that is `parent(p)`. The specific parameters needed vary depending on the type,
for more details about how `TypeParams` are used during (de)serialization see the [developer documentation](@ref dev_serialization).

> Since it can be useful for users to create them when overriding parameters on load we include a small section here.

Reviewer: > [...] we include a small section here.

Does this refer to the section ending here? If so then I object to this sentence, because this section does not actually tell me how to "create them" nor how to use them to "override parameters" during load...

Author: Ok, maybe there is a sentence missing to please check the docstring for type_params below?

> Since it can be useful for users to create them when overriding parameters on load we include a small section here.

Author: ```suggestion
Since it can be useful for users to create them when overriding parameters on load we include some documentation of using the `type_params` function below.
```

```@docs
type_params
```

## Container Types

> We use the name container types to refer to the types `Vector`, `Tuple`, `NamedTuple`, `Set`, `Dict` and `Matrix`. 

Reviewer: ```suggestion
We use the term "container types" to refer to the types `Vector`, `Tuple`, `NamedTuple`, `Set`, `Dict` and `Matrix`. 
```

> Serializing container types that contain serializable OSCAR types is also possible.

Reviewer: The "also" in this sentence seems out of place? Maybe
```suggestion
Container types that contain serializable OSCAR types can themselves be serialized.
```

However, there are some caveats. Although your data type might be a `Vector` of an OSCAR type say `Vector{MPolyRingElem}`, this type might not be serializable.
To serialize a `Vector` all entries must have the output of [`type_params`](@ref) be equal (`==` returns true), this allows us to provide a size efficient storage of homogeneous data. The `Tuple` and `NamedTuple` types will store type information (type names and parameters) for each entry separately, this is the recommended way of serializing collections of types
with different `TypeParams`. 
Storing `Dict` types depends on the key and value types, for example if the keys are either `String`, `Symbol` or `Int` then the type parameters for the value may vary by key. 

> In the other cases the type parameters for all keys must be equal, and likewise all values must have equal type parameters."

Reviewer: ```suggestion
In the other cases the type parameters for all keys must be equal, and likewise all values must have equal type parameters.
```

> ```jldoctest; setup=:(current=pwd(); cd(mktempdir()), teardown=:(cd(current))

Reviewer 2: ```suggestion
```jldoctest; setup=:(current=pwd(); cd(mktempdir())), teardown=:(cd(current))
```
